src/NuGet.Core/NuGet.Packaging/Signing/Archive/ReadOnlyBufferedStream.cs

```diff
@@ -0,0 +1,365 @@
+        {
+            ThrowIfDisposed();
+
+            return base.CopyToAsync(destination, bufferSize, cancellationToken);
+        }
+
+        public override void Flush()
+        {
+            ThrowIfDisposed();
+
+            // This is a read-only stream.
+            throw new NotSupportedException();
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            // This is a read-only stream.
+            throw new NotSupportedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            ThrowIfDisposed();
+
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (count - offset > buffer.Length ||
+                count + offset > buffer.Length)
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.RangeOutOfBoundsForArray,
+                        nameof(offset),
+                        nameof(count)));
+            }
+
+            var bytesRead = 0;
+            var destinationOffset = offset;
+            var bytesWanted = count;
+
+            while (bytesWanted > 0 && !IsPositionAfterEndOfStream())
+            {
+                int sourceStart;
+                int bytesToCopy;
+
+                if (IsPositionInBuffer())
+                {
+                    sourceStart = (int)(Position - _bufferStartPosition);
+                }
+                else
+                {
+                    FillBuffer();
+
+                    sourceStart = 0;
+                }
+
+                bytesToCopy = (int)Math.Min(_bufferStartPosition + _bufferFillLength - Position, bytesWanted);
+
+                Debug.Assert(bytesToCopy >= 0);
+
+                if (bytesToCopy <= 0)
+                {
+                    break;
+                }
+
+                Buffer.BlockCopy(_buffer, sourceStart, buffer, destinationOffset, bytesToCopy);
+
+                destinationOffset += bytesToCopy;
+                bytesRead += bytesToCopy;
+                bytesWanted -= bytesToCopy;
+
+                Position += bytesToCopy;
+            }
+
+            return bytesRead;
+        }
+
+        public override int ReadByte()
+        {
+            ThrowIfDisposed();
+
+            int offset;
+
+            if (IsPositionInBuffer())
+            {
+                offset = (int)(Position - _bufferStartPosition);
```

Comment:
given that Position is long , could this result in overflow?

Reply:
It's not possible.  `IsPositionInBuffer()` guarantees that `Position` is within the range of the stream that buffer represents.  It is this guarantee that ensures this math won't overflow.  `offset` will always be in the range of `[0, _buffer.Length)`.